File: day-02/password_philosophy.go
package main

import (
	"fmt"
	"io/ioutil"
	"regexp"
	"strconv"
	"strings"
)

const input_file string = "./day-02/input.txt"

type is_valid func(int, int, byte, string) bool

func CountValidPasswords(input string, valid_fn is_valid) int {
	totalValid := 0
	lines := strings.Split(strings.TrimSpace(input), "\n")
	lineMatcher := regexp.MustCompile(`^(\d+)-(\d+) (\w): (\w+)$`)
	for _, line := range lines {
		matches := lineMatcher.FindStringSubmatch(line)
		min, err := strconv.Atoi(matches[1])
		if err != nil {
			panic(err)
		}
		max, err := strconv.Atoi(matches[2])
		if err != nil {
			panic(err)
		}
		if valid_fn(min, max, matches[3][0], matches[4]) {
			totalValid++
		}
	}
	return totalValid
}

func IsValidPart1(min int, max int, char byte, pattern string) bool {
	charCount := 0
	for i := range pattern {
		if pattern[i] == char {
			charCount++
		}
	}
	return charCount >= min && charCount <= max
}

func Part1(input string) int {
	return CountValidPasswords(input, IsValidPart1)
}

func IsValidPart2(min int, max int, char byte, pattern string) bool {
	return (pattern[min-1] == char) != (pattern[max-1] == char)
}

func Part2(input string) int {
	return CountValidPasswords(input, IsValidPart2)
}

func main() {
	b, err := ioutil.ReadFile(input_file)
	if err != nil {
		fmt.Print(err)
	}
	data := string(b)

	fmt.Printf("Part 1: %d\n", Part1(data))
	fmt.Printf("Part 2: %d\n", Part2(data))
}

File: day-02/password_philosophy.py
import re

input_file = 'day-02/input.txt'

def count_valid_passwords(input, is_valid):
  total_valid = 0
  for line in input.splitlines():
    min, max, char, pattern = list(re.match(r"(\d+)-(\d+) (\w): (\w+)", line).groups())
    if is_valid(pattern, char, int(min), int(max)):
      total_valid += 1
  return total_valid

def part1(input):
  is_valid = lambda pattern, char, min, max: min <= pattern.count(char) <= max
  return count_valid_passwords(input, is_valid)

def part2(input):
  is_valid = lambda pattern, char, min, max: (pattern[min-1] == char) ^ (pattern[max-1] == char) # Not zero indexed
  return count_valid_passwords(input, is_valid)

if __name__ == "__main__":
  with open(input_file) as f:
    data = f.read()
  print("Part 1: ", part1(data))
  print("Part 2: ", part2(data))